File: mirascope/llm/calls/call.py
"""The Call module for generating responses using LLMs."""

from dataclasses import dataclass
from typing import ParamSpec

from ..prompts import Promptable
from ..responses import AsyncStream, Response, Stream
from .base_call import BaseCall

P = ParamSpec("P")


@dataclass
class Call(BaseCall[P, Promptable]):
    """A class for generating responses using LLMs."""

    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> Response:
        """Generates a response using the LLM."""
        raise NotImplementedError()

    async def call_async(self, *args: P.args, **kwargs: P.kwargs) -> Response:
        """Generates an asynchronous response using the LLM."""
        raise NotImplementedError()

    def stream(self, *args: P.args, **kwargs: P.kwargs) -> Stream:
        """Generates a streaming response using the LLM."""
        raise NotImplementedError()

    async def stream_async(self, *args: P.args, **kwargs: P.kwargs) -> AsyncStream:
        """Generates an asynchronous streaming response using the LLM."""
        raise NotImplementedError()

File: typescript/mirascope/src/llm/calls/call.ts
/**
 * @fileoverview The Call module for generating responses using LLMs.
 */

import type { Prompt } from "../prompts";
import type { Response } from "../responses";
import { BaseCall } from "./base-call";

// TODO: figure out named call parameters

/**
 * A class for generating responses using LLMs.
 */
export class Call<P extends any[] = any[]> extends BaseCall<P, Prompt<P>> {
  /**
   * Generates a response using the LLM.
   */
  call(...args: P): Response {
    throw new Error("Not implemented");
  }

  /**
   * Generates an asynchronous response using the LLM.
   */
  async callAsync(...args: P): Promise<Response> {
    throw new Error("Not implemented");
  }
}
